iree/turbine/kernel/wave/barriers.py

```diff
@@ -49,10 +52,9 @@ def add_shared_memory_barriers(
             )
 
     # Synchronize before the write to shared memory to avoid stepping over
-    # reads in the previous iteration of a loop.
-    if is_reduction_subgraph(graph) and last_node:
-        # Insert barrier at start of block, if load from shared memory exist.
-        with graph.inserting_after(graph._root):
-            SharedMemoryBarrier().add_to_graph(graph)
+    # shared reads in the previous iteration of a loop.
+    if is_reduction_subgraph(graph) and last_node and not check_cycle:
+        # Add barriers between ops from different iterations in the same loop.
+        add_shared_memory_barriers(trace, graph, last_node, check_cycle=True)
```

Comment:
Why do we need check_cycle? Doesnt seem to be used anywhere above?

Reply:
discussed offline, it was to check for next iteration, renamed to make it more intuitive :)